File: generate-feed-science.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET
from datetime import datetime, timezone
from pathlib import Path

OUT = Path("feed.json")
MAX_ITEMS = 200

QUERIES = [
    ("Superconductivity", "all:superconductor OR all:superconductivity OR cat:cond-mat.supr-con"),
    ("Quantum", 'cat:quant-ph OR all:"quantum computing" OR all:"quantum information"'),
    ("Tribology", "all:tribology OR all:friction OR all:wear OR all:lubrication"),
    ("Materials science", "cat:cond-mat.mtrl-sci OR all:polymer OR all:nanomaterial OR all:materials"),
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; ScienceFeed/1.0; +https://github.com/tasselrex/science)"
}

def clean(s: str | None) -> str:
    if not s:
        return ""
    return re.sub(r"\s+", " ", s).strip()

def strip_html(s: str | None) -> str:
    if not s:
        return ""
    s = re.sub(r"<[^>]+>", " ", s)
    s = clean(s)
    return (
        s.replace("&amp;", "&")
         .replace("&lt;", "<")
         .replace("&gt;", ">")
         .replace("&quot;", '"')
         .replace("&apos;", "'")
    )

def truncate(s: str, n: int = 180) -> str:
    s = clean(s)
    return s if len(s) <= n else s[: n - 1].rstrip() + "…"

def fetch(url: str, timeout: int = 25) -> str:
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="replace")

def parse_atom(xml_text: str, source_label: str):
    ns = {"a": "http://www.w3.org/2005/Atom"}
    root = ET.fromstring(xml_text)
    items = []

    for entry in root.findall("a:entry", ns):
        title = clean(entry.findtext("a:title", default="", namespaces=ns))
        summary = truncate(strip_html(entry.findtext("a:summary", default="", namespaces=ns)), 180)
        published = clean(entry.findtext("a:published", default="", namespaces=ns))[:10]
        updated = clean(entry.findtext("a:updated", default="", namespaces=ns))[:10]

        link = ""
        for link_el in entry.findall("a:link", ns):
            if link_el.attrib.get("rel", "alternate") == "alternate" and link_el.attrib.get("href"):
                link = link_el.attrib["href"]
                break
        if not link:
            link = clean(entry.findtext("a:id", default="", namespaces=ns))

        authors = [
            clean(author.findtext("a:name", default="", namespaces=ns))
            for author in entry.findall("a:author", ns)
        ]
        authors = ", ".join([a for a in authors if a][:3])

        items.append({
            "title": title,
            "summary": summary,
            "source": source_label,
            "date": published or updated or datetime.now(timezone.utc).date().isoformat(),
            "link": link,
            "authors": authors,
        })

    return items

def arxiv_url(query: str) -> str:
    base = "https://export.arxiv.org/api/query"
    params = {
        "search_query": query,
        "start": "0",
        "max_results": "50",
        "sortBy": "submittedDate",
        "sortOrder": "descending",
    }
    return base + "?" + urllib.parse.urlencode(params)

def main():
    print("Starting feed generation", flush=True)

    buckets = {label: [] for label, _ in QUERIES}
    errors = []
    source_counts = {}

    per_source_limit = max(1, MAX_ITEMS // max(1, len(QUERIES)))

    for label, query in QUERIES:
        try:
            print("Fetching:", label, flush=True)
            xml_text = fetch(arxiv_url(query))
            print("Fetched:", label, flush=True)

            items = parse_atom(xml_text, label)
            print(label, len(items), items[0]["title"] if items else "NO HITS", flush=True)

            seen_titles = set()
            unique_items = []
            for item in items:
                key = item["title"].lower()
                if key in seen_titles:
                    continue
                seen_titles.add(key)
                unique_items.append(item)

            buckets[label] = unique_items[:per_source_limit]
            source_counts[label] = len(buckets[label])

        except Exception as e:
            print("FAILED:", label, e, flush=True)
            errors.append(f"{label}: {e}")
            source_counts[label] = 0

    final_items = []
    global_seen = set()

    while len(final_items) < MAX_ITEMS:
        made_progress = False

        for label, _ in QUERIES:
            bucket = buckets[label]
            while bucket:
                item = bucket.pop(0)
                key = item["title"].lower()
                if key in global_seen:
                    continue
                global_seen.add(key)
                final_items.append(item)
                made_progress = True
                break

            if len(final_items) >= MAX_ITEMS:
                break

        if not made_progress:
            break

    payload = {
        "updatedAt": datetime.now(timezone.utc).isoformat(),
        "items": final_items,
        "errors": errors,
        "sourceCounts": source_counts,
        "sourceTotal": len(QUERIES),
        "itemTotal": len(final_items),
    }

    OUT.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote {OUT} with {len(payload['items'])} items", flush=True)
    print("Source counts:", source_counts, flush=True)

if __name__ == "__main__":
    main()

File: generate-feed-science.mjs
#!/usr/bin/env node
/**
 * Generates feed.json for a consciousness studies feed.
 * Sources: arXiv API (official Atom feed format).
 *
 * No external dependencies required.
 */

const fs = require("fs/promises");

const OUT_FILE = "feed.json";
const MAX_ITEMS = 24;

// Tune these queries as needed.
const SEARCHES = [
  {
    source: "AI / machine learning",
    query: 'all:"machine learning" OR all:"deep learning" OR all:"large language model" OR all:"foundation model"'
  },
  {
    source: "Biology / biotech",
    query: 'all:biolog* OR all:protein OR all:antibody OR all:"cell therapy" OR all:"gene editing"'
  },
  {
    source: "Microscopy / imaging",
    query: 'all:microscopy OR all:"image analysis" OR all:"cryo-EM" OR all:"electron microscopy" OR all:imaging'
  },
  {
    source: "Materials science",
    query: 'all:"materials science" OR all:polymer OR all:nanomaterial OR all:degradation OR all:surface'
  },
  {
    source: "Analytical chemistry",
    query: 'all:"mass spectrometry" OR all:chromatography OR all:spectroscopy OR all:FTIR OR all:Raman'
  },
  {
    source: "Regulatory / pharma",
    query: 'all:FDA OR all:USP OR all:EMA OR all:"good manufacturing practice" OR all:pharma'
  }
];

function escapeXml(s = "") {
  return String(s)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

function stripTags(s = "") {
  return String(s).replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim();
}

function htmlToText(s = "") {
  return stripTags(s)
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'");
}

function truncate(s, n) {
  const text = String(s || "").trim();
  if (text.length <= n) return text;
  return text.slice(0, Math.max(0, n - 1)).trimEnd() + "…";
}

function parseArxivFeed(xml, sourceLabel) {
  const entries = [];
  const entryMatches = xml.match(/<entry>[\s\S]*?<\/entry>/g) || [];

  for (const entry of entryMatches) {
    const get = (tag) => {
      const m = entry.match(new RegExp(`<${tag}>([\\s\\S]*?)<\\/${tag}>`));
      return m ? htmlToText(m[1]) : "";
    };

    const title = get("title").replace(/\s+/g, " ").trim();
    const summary = get("summary").replace(/\s+/g, " ").trim();
    const published = get("published").slice(0, 10);
    const updated = get("updated").slice(0, 10);

    const idMatch = entry.match(/<id>([\s\S]*?)<\/id>/);
    const linkMatch = entry.match(/<link[^>]*rel="alternate"[^>]*href="([^"]+)"/) || entry.match(/<id>([\s\S]*?)<\/id>/);

    const authors = [...entry.matchAll(/<author>[\s\S]*?<name>([\s\S]*?)<\/name>[\s\S]*?<\/author>/g)]
      .map(m => htmlToText(m[1]))
      .filter(Boolean)
      .slice(0, 3)
      .join(", ");

    entries.push({
      title,
      summary: truncate(summary, 180),
      source: sourceLabel,
      date: published || updated || new Date().toISOString().slice(0, 10),
      link: linkMatch ? htmlToText(linkMatch[1] || linkMatch[0]) : (idMatch ? htmlToText(idMatch[1]) : "https://arxiv.org/"),
      authors
    });
  }

  return entries;
}

async function fetchWithTimeout(url, timeoutMs = 20000) {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; ConsciousnessFeed/1.0; +https://github.com/tasselrex/AIResearchFeed)"
      }
    });
    if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
    return await res.text();
  } finally {
    clearTimeout(t);
  }
}

async function gather() {
  const all = [];
  for (const s of SEARCHES) {
    const url = new URL("https://export.arxiv.org/api/query");
    url.searchParams.set("search_query", s.query);
    url.searchParams.set("start", "0");
    url.searchParams.set("max_results", "8");
    url.searchParams.set("sortBy", "submittedDate");
    url.searchParams.set("sortOrder", "descending");

    try {
      const xml = await fetchWithTimeout(url.toString());
      const items = parseArxivFeed(xml, s.source);
      all.push(...items);
    } catch (err) {
      console.error(`Failed to fetch ${s.source}:`, err.message);
    }
  }

  // Deduplicate by title and keep the newest, most relevant ones.
  const seen = new Set();
  const deduped = [];
  for (const item of all) {
    const key = item.title.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(item);
  }

  // Prefer newer items; keep stable order within same date.
  deduped.sort((a, b) => String(b.date).localeCompare(String(a.date)));

  const payload = {
    updatedAt: new Date().toISOString(),
    items: deduped.slice(0, MAX_ITEMS)
  };

  await fs.writeFile(OUT_FILE, JSON.stringify(payload, null, 2) + "\n", "utf8");
  console.log(`Wrote ${OUT_FILE} with ${payload.items.length} items`);
}

gather().catch(err => {
  console.error(err);
  process.exit(1);
});
